plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -277,7 +304,9 @@ public static ReaderPageSource createPageSource(
                     }));
         }
 
-        ConnectorPageSource parquetPageSource = new ParquetPageSource(parquetReader, trinoTypes.build(), internalFields.build());
+        ConnectorPageSource parquetPageSource = columns.contains(ROW_INDEX_COLUMN)
+                ? new RowIndexPageSource(parquetReader, trinoTypes.build(), internalFields.build(), columns.indexOf(ROW_INDEX_COLUMN))
+                : new ParquetPageSource(parquetReader, trinoTypes.build(), internalFields.build());
```

Comment:
I'd consider fusing `RowIndexPageSource` with `ParquetPageSource`, i.e.
handle injection of the rowindex within https://github.com/trinodb/trino/blob/fe65c836fc2803c2cb04457c21e73c39b9f9b55f/plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java#L93-L103 loop

this would avoid copying Block[] and shouldn't make that class more complex

Reply:
I like having it as a separate class because it feels much cleaner than having more field-based behavior switches, but I'm leaning the other way now because it would simplify the factory.